code/game/machinery/atmoalter/gas_mine.dm

```diff
@@ -7,29 +9,81 @@
 
 	starting_materials = null
 	w_type = NOT_RECYCLABLE
+	var/power_load = 0
+	var/rate							//moles generated last tick. used when examining
+	var/moles_outputted					//moles outputted last tick. used when examining
+	var/base_gas_production = 4500		//base KPa per tick - without external power
+	var/max_external_pressure = 10000	//max KPa output - without external power
+	var/on = TRUE
 
-	var/datum/gas_mixture/air_contents
-	var/datum/gas_mixture/pumping = new //used in transfering air around
-
-	var/on=1
-
-	var/max_external_pressure=10000 // 10,000kPa ought to do it.
-	var/internal_pressure=4500 // Bottleneck
+	var/list/gases = list()				//which gases the miner generates
+	var/datum/gas_mixture/air_contents	//which gases the miner generates, and how fast (in KPa per tick)
+	var/datum/gas_mixture/pumping 		//used in transfering air around
 
 	var/overlay_color = "#FFFFFF"
 
 	machine_flags = WRENCHMOVE | FIXED2WORK
 
 /obj/machinery/atmospherics/miner/New()
 	..()
+	
+	pumping = new
 	air_contents = new
 	air_contents.volume = 1000
 	pumping.volume = 1000 //Same as above so copying works correctly
 	air_contents.temperature = T20C
-	AddAir()
-	air_contents.update_values()
+	set_rate(base_gas_production)
 	update_icon()
 
+/obj/machinery/atmospherics/miner/Destroy()
+	if(pumping)
+		qdel(pumping)
+		pumping = null
+	if(air_contents)
+		qdel(air_contents)
+		air_contents = null
+	..()
+
+
+//update gas creation speed into air_contents
+/obj/machinery/atmospherics/miner/proc/set_rate(var/internal_pressure)
+	air_contents.remove(air_contents.total_moles)//set to 0
+	//rate is in mols
+	rate = internal_pressure * air_contents.volume / (R_IDEAL_GAS_EQUATION * air_contents.temperature)
+	//this is ugly
```

Comment:
Replace this with `for(var/i = 1, i <= gases.len, i++)
		air_contents.adjust_gas(gases[i], gases[gases[i]]*rate)`

Reply:
this is referring to the part that isn't quoted there, the giant blocks of adjust_multi, right?